kong/db/dao/init.lua

```diff
@@ -1479,28 +1498,38 @@ function DAO:cache_key(key, arg2, arg3, arg4, arg5, ws_id)
     error("key must be a string or an entity table", 2)
   end
 
-  if key.ws_id then
+  if key.ws_id ~= nil and key.ws_id ~= null then
     ws_id = key.ws_id
   end
 
   local values = new_tab(7, 0)
-  values[1] = self.schema.name
-  local source = self.schema.cache_key or self.schema.primary_key
+  values[1] = name
 
   local i = 2
-  for j = 1, #source do
-    local name = source[j]
-    local field = self.schema.fields[name]
-    local value = key[name]
-    if value == null or value == nil then
-      value = ""
-    elseif field.type == "foreign" then
-      -- FIXME extract foreign key, do not assume `id`
-      value = value.id
+
+  local fields = schema.fields
+  local source = schema.cache_key
+  local use_pk = true
+  if source then
+    for j = 1, #source do
+      local value = get_cache_key_value(source[j], key, fields)
+      if value ~= nil then
+        use_pk = false
+      end
+      values[i] = value or ""
```

Comment:
Perhaps an extremely edge case, but if an entity has a cache key value which includes a `boolean`, the `false`s will get transformed into "" here. Also I am not sure how safe it is to remove the `tostring`
```suggestion
      if value == nil then
        value = ""
      else
        use_pk = false
      end
      values[i] = tostring(value)
```

Reply:
AFAIK this is not needed as the `value` comes from `get_cache_key_value` that can only return `nil` or a `string` (never a `boolean` for example).